src/main/java/org/craftercms/engine/util/servlet/ConfigAwareProxyServlet.java

```diff
@@ -84,4 +110,105 @@ protected void copyResponseHeaders(HttpResponse proxyResponse, HttpServletReques
         }
     }
 
+    @Override
+    protected void service(HttpServletRequest servletRequest, HttpServletResponse servletResponse)
+            throws IOException, ServletException {
+        var websocketKey = servletRequest.getHeader(WEBSOCKET_IDENTIFIER_HEADER);
+        if (websocketKey != null) {
+            websocketHandlerService(servletRequest, servletResponse);
+        } else {
+            super.service(servletRequest, servletResponse);
+        }
+    }
+
+    /**
+     * Handle a websocket request by upgrading requests with 101 responses
+     * @param servletRequest request
+     * @param servletResponse response
+     */
+    private void websocketHandlerService(HttpServletRequest servletRequest, HttpServletResponse servletResponse) throws IOException, ServletException {
+        //initialize request attributes from caches if unset by a subclass by this point
+        if (servletRequest.getAttribute(ATTR_TARGET_URI) == null) {
+            servletRequest.setAttribute(ATTR_TARGET_URI, targetUri);
+        }
+        if (servletRequest.getAttribute(ATTR_TARGET_HOST) == null) {
+            servletRequest.setAttribute(ATTR_TARGET_HOST, targetHost);
+        }
+        String proxyRequestUri = rewriteUrlFromRequest(servletRequest);
+        URL url = new URL(proxyRequestUri);
+
+        Socket sock = new Socket(url.getHost(), url.getPort());
+        boolean closeSocket = false;
+        try {
+            var sockIn = sock.getInputStream();
+            var sockOut = sock.getOutputStream();
+
+            StringBuilder req = new StringBuilder(512);
```

Comment:
Isn't there a library that already does this? I mean, that basically can do WS clients? Seems like wer'e doing everything from scratch